src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+
+        # write BAM file
+        with pysam.AlignmentFile(os.path.join(output_path, self.bam_output_filename),
+                                 mode='wb',
+                                 reference_names=[SyntheticDataBundle.synthetic_sequence_name],
+                                 reference_lengths=[SyntheticDataBundle.synthetic_sequence_length]) as bo:
+            for record in records:
+                bo.write(record)
+
+        # write count matrix, row index, and col index
+        np.save(os.path.join(output_path, self.count_matrix_output_filename), synthetic_data_bundle.count_matrix)
+        np.save(os.path.join(output_path, self.row_index_output_filename), synthetic_data_bundle.row_index)
+        np.save(os.path.join(output_path, self.col_index_output_filename), synthetic_data_bundle.col_index)
+
+    def _generate_random_count_matrix(self) -> np.ndarray:
+        non_zero_count_matrix = self.rng.poisson(
+            lam=self.gene_expression_rate, size=(self.num_cells, self.max_genes))
+        count_matrix = np.zeros((self.num_cells, self.num_genes), dtype=np.int)
+        for i, i_gene in enumerate(self.to_be_used_gene_indices):
+            count_matrix[:, i_gene] = non_zero_count_matrix[:, i]
+        return count_matrix
+
+    def _generate_primary_alignment_record_bundle(self, count_matrix: np.ndarray) \
+            -> Tuple[Set[AlignmentRecordTags], List[str], List[str]]:
+        alignments: Set[AlignmentRecordTags] = set()
+        used_cell_barcodes: Set[str] = set()
+
+        row_index: List[str] = []
+        col_index = self.all_gene_names
+
+        for i_cell in range(self.num_cells):
+            # generate a unique cell barcode
+            while True:
+                cell_barcode = self._generate_random_cell_barcode()
+                if cell_barcode not in used_cell_barcodes:
+                    break
+            row_index.append(cell_barcode)
+
+            for i_gene in self.to_be_used_gene_indices:
+                for i_molecule in range(count_matrix[i_cell, i_gene]):
+                    # generate a unique alignment tag
+                    unique_alignment_tag = self._generate_unique_random_alignment_tag(
+                        alignments,
+                        gene_name=self.all_gene_names[i_gene],
+                        cell_barcode=cell_barcode)
+                    alignments.add(unique_alignment_tag)
+
+        return alignments, row_index, col_index
+
+    def _generate_unique_random_alignment_tag(self,
+                                              existing_alignment_tags: Set[AlignmentRecordTags],
+                                              gene_name: str,
+                                              cell_barcode: Optional[str] = None,
+                                              molecule_barcode: Optional[str] = None) -> AlignmentRecordTags:
+        assert gene_name in self.to_be_used_gene_names, f"{gene_name} is not an allowed gene for generating " \
```

Comment:
I think I'd put the slash here before the string, to give more room to expand it in the future.

Reply:
Done.